src/attribute_collector.cpp

```diff
@@ -0,0 +1,105 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025 Datadog, Inc.
+
+#include <span>
+#include <string>
+#include <string_view>
+#include <unordered_map>
+
+#include "attribute_collector.hpp"
+#include "ddwaf.h"
+#include "object_store.hpp"
+#include "target_address.hpp"
+#include "utils.hpp"
+
+namespace ddwaf {
+
+bool attribute_collector::emplace(std::string_view key, const ddwaf_object &value, bool copy)
+{
+    if (emplaced_or_pending_attributes_.contains(key)) {
+        return false;
+    }
+
+    return emplace_helper(key, value, copy);
+}
+
+bool attribute_collector::collect(const object_store &store, target_index input_target,
+    std::span<const std::string> input_key_path, std::string_view output)
+{
+    if (emplaced_or_pending_attributes_.contains(output)) {
+        return false;
+    }
+
+    auto state = collect_helper(store, input_target, input_key_path, output);
+    if (state == collection_state::unavailable) {
+        pending_.emplace(output, target_type{input_target, input_key_path});
+    }
+    return state != collection_state::failed;
+}
+
+void attribute_collector::collect_pending(const object_store &store)
+{
+    for (auto it = pending_.begin(); it != pending_.end();) {
+        auto output = it->first;
+        // No need to check if the key is already present, as emplace and collect
+        // already perform the check. As for items in the pending map, since the
+        // map key is the output key, duplicates aren't possible.
+        auto &[input_target, input_key_path] = it->second;
+
+        auto state = collect_helper(store, input_target, input_key_path, output);
+        if (state != collection_state::unavailable) {
+            it = pending_.erase(it);
+        } else {
+            ++it;
+        }
+    }
+}
+
+attribute_collector::collection_state attribute_collector::collect_helper(const object_store &store,
+    target_index input_target, std::span<const std::string> input_key_path, std::string_view output)
```

Comment:
maybe some logging here

Reply:
True I should add a bit more logging all around.

Reply:
Added some logging.